industry_benchmarks/utils/tape_networks.py

```diff
@@ -0,0 +1,459 @@
+                edges[name] = t
+
+    # get only finished edges:
+    fin_edge = []
+    for name in edges:
+        edge = edges[name]
+        ec = edge_count[name]
+
+        if ec["solvent"] >= 1 and ec["protein"] >= 1:
+            fin_edge.append(edge)
+
+    alchemical_network = AlchemicalNetwork(nodes=nodes, edges=fin_edge)
+    return alchemical_network
+
+
+def alchemical_network_to_ligand_network(alchemical_network) -> LigandNetwork:
+    edges = []
+    nodes = []
+    for e in alchemical_network.edges:
+        nA = e.stateA.components["ligand"]
+        nB = e.stateB.components["ligand"]
+
+        edges.append(e.mapping)
+        nodes.append(nA)
+        nodes.append(nB)
+
+    network = LigandNetwork(nodes=set(nodes), edges=set(edges))
+    return network
+
+
+def ligand_network_to_networkx(ligand_network) -> nx.Graph:
+
+    edges = ligand_network.edges
+    wedges = []
+    for edge in list(edges):
+        wedges.append(
+            [edge.componentA, edge.componentB, edge.annotations["score"]]
+        )
+
+    g = nx.Graph(nodes=ligand_network.nodes)
+    g.add_weighted_edges_from(ebunch_to_add=wedges)
+    return g
+
+
+
+def decomposite_disconnected_ligand_network(
+    ligand_network: LigandNetwork,
+) -> list[LigandNetwork]:
+    # build alchemical sub-networks
+    g = ligand_network_to_networkx(ligand_network)
+    all_edges = list(ligand_network.edges)
+    sub_network_nodes = list(nx.connected_components(g))
+
+    ligand_sub_networks = []
+    for i, snn in enumerate(sub_network_nodes):
+        # collect edges
+        sub_edges = []
+        for e in all_edges:
+            if any(n in (e.componentA, e.componentB) for n in snn):
+                sub_edges.append(e)
+
+        ligand_sub_network = LigandNetwork(nodes=snn, edges=sub_edges)
+        ligand_sub_networks.append(ligand_sub_network)
+
+    return ligand_sub_networks
+
+
+def get_new_network_tapes(
+    ligand_sub_networks: list[LigandNetwork], 
+    input_ligand_network: LigandNetwork,
+    n_connecting_edges: int = 3
+
+) -> LigandNetwork:
+
+    mapper = KartografAtomMapper()
+    scorer = lomap_scorers.default_lomap_score
+    concatenator = MstConcatenator(
+        mapper=mapper,
+        scorer=scorer,
+        n_connecting_edges=n_connecting_edges,
+        n_processes=1,
+    )
+
+    in_edges = input_ligand_network.edges
+
+    concatenated_network = ligand_sub_networks[0]
+    tape_edges = []
+    for ligand_sub_network in ligand_sub_networks[1:]:
+        nedges= min(n_connecting_edges*10, len(concatenated_network.edges))
+        
+        concatenator.n_connecting_edges = nedges
+        concatenated_network = concatenate_networks(
+            [ligand_sub_network, concatenated_network],
+            concatenator=concatenator,
+        )
+        concatenated_edges = concatenated_network.edges
+        possible_edges = list(sorted(list(concatenated_edges.difference(in_edges)), key=lambda e: e.annotations["score"], reverse=True))
```

Comment:
In my case it now still wrote out edges that are the reverse of the transformation from before, so if A-->B had failed, it will still write out a new B-->A transformation. Maybe you could add the flipped edges here as well to the `in_edges`?

Reply:
oooh good catch!

Reply:
I think it is fixed now.